### PR TITLE
Changes to read from BigQuery 

### PR DESCRIPTION
**Pull request by LAKSHMIRPILLAI**
## What does this change?

Bandit currently uses Athena to query for impressions and annualised value per test variant. This is currently for epics only as we do not have banner impressions data in AWS.But all the required data for Bandit is available in BigQuery tables.So we can use BigQuery.

To get the
views related data - we use the online_traffic.fact_page_view table
acquisitions related data - we use the datalake.fact_acquisition_event table
and for currency conversion rates we use the datalake.fixer_exchange_rates table

Also in this PR, we are implementing a simple annualised value calculator .

This PR only covers the changes for the Epic bandit tests.We will have follow up PR with the step function modification  for the Banner and Live blog Epic tests.

[Trello card](https://trello.com/c/ilGhIbCW/597-migrate-bandits-to-use-bigquery)

##How to test

As we did not have. enough data in the CODE tables, we pointed the bandit system to use the PROD tables for querying in the [PR](https://github.com/guardian/gcp-iac-terraform/pull/885)

Then we recreated a test from RRCP Prod env  in RRCP CODE  with the same name (that has sufficient data in the online_traffic.fact_page_view table  and atalake.fact_acquisition_event table, need not be a bandit test).
Mark the test in CODE RRCP as a bandit test.

eg: we used 2024_08_30_CHOICE_THEORY_EPIC__UK

On starting a new execution of the step function, we received the new bandit data in the Dynamo DB table for the test

image

Also on querying the PROD database we received the  below 

image

### CODE REVIEW

File: bandit/src/lib/query.ts [hunk 1]
```diff
@@ -1,11 +1,3 @@
-import type * as AWS from 'aws-sdk';
-import type {
-	GetQueryExecutionOutput, GetQueryResultsOutput,
-	QueryExecutionId, QueryExecutionState,
-	StartQueryExecutionInput,
-} from "aws-sdk/clients/athena";
-import {QueryExecutionNotFoundError, QueryFailedError, QueryPendingError} from "./errors";
-
 export class Query {
```
Comment (tomrf1) [inline bandit/src/lib/query.ts:1]: This model isn't necessary anymore, I think we needed the token for athena.
When we call `query()` on the BigQuery client we can actually just pass in a string

File: bandit/src/query-lambda/queryBigQueryData.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
+import {format, subDays} from "date-fns";
```
Comment (tomrf1) [inline bandit/src/query-lambda/queryBigQueryData.ts:1]: Other filenames in this repo are hyphenated.
Also, the function in this file just builds the query, whereas I think the filename suggests it runs the query

File: bandit/src/query-lambda/queryBigQueryData.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+import type {Test} from "../lib/models";
+import {Query} from "../lib/query";
+
+
+const ANNUALISED_VALUE_CAP = 250;
+export const buildQuery = (
+	test: Test,
+	stage: "CODE" | "PROD",
```
Comment (tomrf1) [inline bandit/src/query-lambda/queryBigQueryData.ts:9]: How about using the `stage` parameter for the query, and outside of this function we can hardcode passing in `PROD`, with a comment explaining why we use it for both environments?
E.g. `FROM datatech-platform-${stage.toLowerCase()}.datalake.fact_acquisition_event`

File: cdk/lib/bandit.ts [hunk 4]
```diff
@@ -103,68 +136,17 @@ export class Bandit extends GuStack {
...
 				getBanditTestsTask
-					.next(queryTask)
-					.next(
-						calculateTask.addRetry({
-							errors: ['QueryPendingError'],
-							interval: Duration.seconds(10),
-							maxAttempts: 5,
-						}),
-					)
+					.next(queryTask.addRetry({
```
Comment (tomrf1) [inline cdk/lib/bandit.ts:145]: I don't think we need a retry anymore. Previously we used retries as a mechanism for polling the athena result.

File: bandit/src/query-lambda/query-lambda.ts [hunk 5]
```diff
@@ -1,41 +1,49 @@
...
+	const resultsFromBigQuery: Array<{testName: string; rows: SimpleQueryRowsResponse}> = await Promise.all(
+		input.tests.map(test => getDataForBanditTest(client, stage, test, input.date))
 	);
 
-	return Promise.all(results);
+	const parsedResults = resultsFromBigQuery.map(({testName, rows}) => {
+		const parsed = parseResultFromBigQuery(rows);
+		return buildWriteRequest(parsed, testName, startTimestamp);
+	});
+	const batches = await Promise.all(parsedResults);
```
Comment (tomrf1) [inline bandit/src/query-lambda/query-lambda.ts:39]: there's no need for the `await Promise.all` here as we already awaited above

File: bandit/src/query-lambda/query-lambda.ts [hunk 6]
```diff
@@ -1,41 +1,49 @@
...
-					testName: test.name,
-					startTimestamp: start.toISOString(),
-				})
-			)
+	const resultsFromBigQuery: Array<{testName: string; rows: SimpleQueryRowsResponse}> = await Promise.all(
+		input.tests.map(test => getDataForBanditTest(client, stage, test, input.date))
 	);
 
-	return Promise.all(results);
+	const parsedResults = resultsFromBigQuery.map(({testName, rows}) => {
```
Comment (tomrf1) [inline bandit/src/query-lambda/query-lambda.ts:35]: A more accurate name here would be `writeRequests`

File: bandit/src/query-lambda/query-lambda.ts (hunk 5 above)
Comment (tomrf1) [reply bandit/src/query-lambda/query-lambda.ts:39]: In fact we don't need to reassign to `batches` at all here. The name batches is a bit misleading anyway, as we're just sending the whole array to dynamo

File: bandit/src/query-lambda/dynamo.ts [hunk 7]
```diff
@@ -35,9 +35,9 @@ function buildDynamoRecord(
 	startTimestamp: string
 ): TestSample {
 	const variants = rows.map((row) => ({
-		variantName: row.variantName,
-		annualisedValueInGBP: row.avGbp,
-		annualisedValueInGBPPerView: row.avGbpPerView,
+		variantName: row.variant_name,
+		annualisedValueInGBP: row.sum_av_eur,
+		annualisedValueInGBPPerView: row.sum_av_eur_per_view,
```
Comment (tomrf1) [inline bandit/src/query-lambda/dynamo.ts:40]: Currently we're using the EUR values in fields with GBP in the name.
How about adjusting the query to do the EUR -> GBP conversion?

This example might be useful: https://github.com/guardian/data-platform-models/blob/25c4bfea2028418dc3800281b7232a2d2d085688/dbt/models/datalake/contributions/contributions.sql#L35
It joins to both the EUR rate and the EUR -> GBP rate

Comment (LAKSHMIRPILLAI) [reply bandit/src/query-lambda/dynamo.ts]: yes, I have updated the query and related files to do the EUR->GBP Conversion

**Review by tomrf1 — APPROVED** [review]
great!

**Pull request merged**